flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/avro/AvroParquetRecordFormat.java

```diff
@@ -0,0 +1,236 @@
+
+        checkArgument(
+                restoredOffset == CheckpointedPosition.NO_OFFSET,
+                "The restoredOffset should always be NO_OFFSET");
+
+        return createReader(config, stream, fileLen, splitEnd);
+    }
+
+    @VisibleForTesting
+    GenericData getDataModel() {
+        Class<E> typeClass = getProducedType().getTypeClass();
+        if (org.apache.avro.specific.SpecificRecordBase.class.isAssignableFrom(typeClass)) {
+            return SpecificData.get();
+        } else if (org.apache.avro.generic.GenericRecord.class.isAssignableFrom(typeClass)) {
+            return GenericData.get();
+        } else {
+            return ReflectData.get();
+        }
+    }
+
+    /** Current version does not support splitting. */
+    @Override
+    public boolean isSplittable() {
+        return false;
+    }
+
+    /**
+     * Gets the type produced by this format. This type will be the type produced by the file source
+     * as a whole.
+     */
+    @Override
+    public TypeInformation<E> getProducedType() {
+        return type;
+    }
+
+    private static void checkNotSplit(long fileLen, long splitEnd) {
+        if (splitEnd != fileLen) {
+            throw new IllegalArgumentException(
+                    String.format(
+                            "Current version of AvroParquetRecordFormat is not splittable, "
+                                    + "but found split end (%d) different from file length (%d)",
+                            splitEnd, fileLen));
+        }
+    }
+
+    /**
+     * {@link StreamFormat.Reader} implementation. Using {@link ParquetReader} internally to read
+     * avro {@link GenericRecord} from parquet {@link InputFile}.
+     */
+    private static class AvroParquetRecordReader<E> implements StreamFormat.Reader<E> {
+
+        private final ParquetReader<E> parquetReader;
+
+        private final long offset;
+        private long skipCount;
+        private final boolean checkpointed;
+
+        private AvroParquetRecordReader(ParquetReader<E> parquetReader) {
+            this(parquetReader, CheckpointedPosition.NO_OFFSET, 0, false);
+        }
+
+        private AvroParquetRecordReader(
+                ParquetReader<E> parquetReader, long offset, long skipCount, boolean checkpointed) {
+            this.parquetReader = parquetReader;
+            this.offset = offset;
+            this.skipCount = skipCount;
+            this.checkpointed = checkpointed;
+        }
+
+        @Nullable
+        @Override
+        public E read() throws IOException {
+            E record = parquetReader.read();
+            incrementPosition();
+            return record;
+        }
+
+        @Override
+        public void close() throws IOException {
+            parquetReader.close();
+        }
+
+        @Nullable
+        @Override
+        public CheckpointedPosition getCheckpointedPosition() {
+            return checkpointed ? new CheckpointedPosition(offset, skipCount) : null;
```

Comment:
I'm not fully satisfied with our ability to recover here. So I looked into what the AvroParquetReader is doing internally:

```
            // Injected
            GenericData model = GenericData.get();
            org.apache.hadoop.conf.Configuration conf = new org.apache.hadoop.conf.Configuration();

            // Low level reader - fetch metadata
            ParquetFileReader reader = null;
            MessageType fileSchema = reader.getFileMetaData().getSchema();
            Map<String, String> metaData = reader.getFileMetaData().getKeyValueMetaData();

            // init Avro specific things
            AvroReadSupport<T> readSupport = new AvroReadSupport<>(model);
            ReadSupport.ReadContext readContext =
                    readSupport.init(
                            new InitContext(
                                  conf,
                                    metaData.entrySet().stream()
                                            .collect(Collectors.toMap(e -> e.getKey(), e -> Collections.singleton(e.getValue()))),
                                    fileSchema));
            RecordMaterializer<T> recordMaterializer = readSupport.prepareForRead(conf, metaData, fileSchema, readContext);
            MessageType requestedSchema = readContext.getRequestedSchema();

            // prepare record reader
            ColumnIOFactory columnIOFactory = new ColumnIOFactory(reader.getFileMetaData().getCreatedBy());
            MessageColumnIO columnIO = columnIOFactory.getColumnIO(requestedSchema, fileSchema, true);

            // for recovery
            while (...) {
              reader.skipNextRowGroup();
            }

            // for reading
            PageReadStore pages;
            for (int block = 0; (pages = reader.readNextRowGroup()) != null; block++) {
                RecordReader<T> recordReader = columnIO.getRecordReader(pages, recordMaterializer);
                for (int i = 0; i < pages.getRowCount(); i++) {
                    T record = recordReader.read();
                    emit record;
                }
            }
```

Here we can easily track the block. Even better most of that snippet is already implemented in `ParquetVectorizedInputFormat`, so we may be able to solve both things at once. WDYT?

Reply:
Thanks for your effort for researching and providing the details code logics. Yes, using the low level API ParquetFileReader is another option I've considered and finally got the feeling that it'd be better to use with `BulkFormat` directly like 'ParquetVectorizedInputFormat' did instead of with `StreamFormat` with the following reasons:

-  the read logic is built in the internal low level class `InternalParquetRecordReader` with package private visibility in parquet-hadoop lib which uses another low level class `ParquetFileReader` internally. This makes the implementation of StreamFormat very complicated. I think the design idea of StreamFormat is to simplify the implementation. They do not seem to work together.

-  `ParquetFileReader`reads data in batch mode, i.e. `PageReadStore pages = reader.readNextFilteredRowGroup();`. If we build these logic into StreamFormat(`AvroParquetRecordFormat` in this case), `AvroParquetRecordFormat` has to take over the role `InternalParquetRecordReader` does, including but not limited to 1. read `PageReadStore` in batch mode; 2. manage `PageReadStore`, i.e. read next page when all records in the current page have been consumed and cache it. 3. manage the read index within the current `PageReadStore` because StreamFormat has its own setting for read size, etc. All of these make `AvroParquetRecordFormat` become the `BulkFormat` instead of `StreamFormat`  

- `StreamFormat` can only be used via `StreamFormatAdapter`, which means everything we will do with the low level APIs for parquet-hadoop lib should have not conflict with the built-in logic provided by `StreamFormatAdapter`.

Now we could see if we build these logics into a `StreamFormat` implementation, i.e. `AvroParquetRecordFormat`, all convenient built-in logic provided by the `StreamFormatAdapter` turns into obstacles. There is also a violation of single responsibility principle, i.e. `AvroParquetRecordFormat`will take some responsibility of `BulkFormat`. I guess this were the reasons why 'ParquetVectorizedInputFormat' implemented `BulkFormat` instead of `StreamFormat`.

In order to build a unified parquet implementation, it makes more sense to consider building these code into a `BulkFormat` implementation class. Speaking of "solve both things at once", since the output data types are different, `RowData` vs. `Avro`, extra converter logic should be introduced into the architecture design. This is beyond the scope of this PR. I would suggest to open another ticket to focus on it. Depending on how complicated the issue will be, a new FLIP might be required.

Current implementation follows the design idea of `StreamFormat` and keep everything on the high level and simple. It is therefore easy to implement and easy for user to understand. It is a good fit for simple use cases. It has no conflict to the other solution based on the code mentioned above teams up with `BulkFormat`.

Reply:
Yes, I think ultimately both `AvroParquetRecordFormat` and `ParquetVectorizedInputFormat` should share a common ancestor and be bulk-based. However, it's strictly speaking a performance improvement and probably improves maintainability in the long run (one base format where all the logic resides). For now, your solution already closes some gaps in the feature matrix and has a good value on its own. I'd pursue the other option later in a separate ticket/PR.